ldb_reader.go

```diff
@@ -38,6 +45,68 @@ var (
 	ErrNoLedgerUpdates      = errors.New("no ledger updates have been received yet")
 )
 
+func newLDBReader(path string) (*LDBReader, error) {
+	db, err := newLDB(path)
+	if err != nil {
+		return nil, err
+	}
+
+	return &LDBReader{Db: db}, nil
+}
+
+func newVersionedLDBReader(dirPath string) (*LDBReader, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	reader := &LDBReader{
+		cancelWatcher: cancel,
+	}
+
+	// To initialize this reader, we must first load an LDB:
+	last, err := lookupLastLDBSync(dirPath)
+	if err != nil {
+		return nil, errors.Wrap(err, "checking last ldb sync")
+	}
+	if last == 0 {
+		return nil, fmt.Errorf("no LDB in path (%s)", dirPath)
+	}
+	err = reader.switchLDB(dirPath, last)
+	if err != nil {
+		return nil, errors.Wrap(err, "switching ldbs")
+	}
+
+	// Then we can defer to the watcher goroutine to swap this
+	// reader LDB if a newer one appears:
+	go reader.watchForLDBs(ctx, dirPath, last)
+
+	return reader, nil
+}
+
+func newLDB(path string) (*sql.DB, error) {
+	_, err := os.Stat(path)
+	switch {
+	case os.IsNotExist(err):
+		return nil, fmt.Errorf("no LDB found at %s", path)
+	case err != nil:
+		return nil, err
+	}
+
+	mode := "ro"
```

Comment:
i'd move this `mode` logic into the else clause below (if !ldbVersioning) for clarity